src/Symfony/Component/Security/Http/EventListener/PasswordMigratingListener.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Symfony\Component\Security\Http\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Security\Core\Encoder\EncoderFactoryInterface;
+use Symfony\Component\Security\Core\User\PasswordUpgraderInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Http\Authenticator\PasswordAuthenticatedInterface;
+use Symfony\Component\Security\Http\Event\VerifyAuthenticatorCredentialsEvent;
+
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ *
+ * @final
+ * @experimental in 5.1
+ */
+class PasswordMigratingListener implements EventSubscriberInterface
+{
+    private $encoderFactory;
+
+    public function __construct(EncoderFactoryInterface $encoderFactory)
+    {
+        $this->encoderFactory = $encoderFactory;
+    }
+
+    public function onCredentialsVerification(VerifyAuthenticatorCredentialsEvent $event): void
+    {
+        if (!$event->areCredentialsValid()) {
+            // Do not migrate password that are not validated
+            return;
+        }
+
+        $authenticator = $event->getAuthenticator();
+        if (!$authenticator instanceof PasswordAuthenticatedInterface || !$authenticator instanceof PasswordUpgraderInterface) {
```

Comment:
I think this is note quite right. Check the old `GuardAuthenticationProvider`. The `PasswordUpgraderInterface` *was* on the "user provider"... but we don't have that concept in the new authenticators (though, maybe we could inject the new `ChainUserProvider` service and find one that supports the class to use here?). 

However we make it work, do we have it implemented correctly for the built-in `PasswordAuthenticatedInterface` authenticators - like `form_login`?